internal/erigon_node/deep_reorg.go

```diff
@@ -0,0 +1,175 @@
+// by using hash, parent hash, and block number
+func (c *NodeClient) FindDeepReorgs(ctx context.Context,
+	writer http.ResponseWriter,
+	template *template.Template,
+	requestChannel chan *internal.NodeRequest) {
+	start := time.Now()
+	var err error
+
+	rc := NewRemoteCursor(c, requestChannel)
+	if err = rc.Init(headersDb, headersTable, nil); err != nil {
+		fmt.Fprintf(writer, "Create remote cursor: %v", err)
+		return
+	}
+
+	total, wrongBlocks, errors := c.findDeepReorgsInternally(ctx, template, rc)
+	for _, err := range errors {
+		if err != nil {
+			fmt.Fprintf(writer, "%v\n", err)
+		}
+	}
+
+	for height, reorgLengthMap := range wrongBlocks {
+		for length, count := range reorgLengthMap {
+			fmt.Fprintf(writer, "Block height: %d, length of reorg: %d, count of deep reorgs: %d \n", height, length, count)
+		}
+	}
+
+	fmt.Fprintf(writer, "Reorg iterator: %d, total scanned %s\n", total, time.Since(start))
+	fmt.Fprintf(writer, "Reorg iterator: %d, wrong blocks\n", len(wrongBlocks))
+}
+
+func (c *NodeClient) findDeepReorgsInternally(ctx context.Context,
+	template *template.Template,
+	rc *RemoteCursor,
+) (uint64, map[uint64]map[uint64]uint64, []error) {
+	// mapping to store information of blocks on particular height
+	deepReorgs := make(map[uint64][]*BlockNode)
+	// store key and value fetched from DB
+	var k []byte
+	var v []byte
+	var err error
+	// stores maximum block height in order to find main chain
+	var max uint64 = 0
+	// store block-height, block-hash, and block's parent hash
+	var height, hash, parentHash uint64
+	// mapping to store block information corresponding to their block hash
+	blocks := make(map[uint64]*BlockNode)
+	// bool to avoid adding 1st node as children to a nil node
+	var isFirst = true
+	var countBlocks uint64 = 0
+	// loop to iterate over all blocks as key value pairs and store their info
+	for k, v, err = rc.Next(); err == nil && k != nil; k, v, err = rc.Next() {
+		// fetch block height, hash, and parentHash from key vlaue pair
+		height = binary.BigEndian.Uint64(k[:8])
+		hash = binary.BigEndian.Uint64(k[8:40])
+		parentHash = binary.BigEndian.Uint64(v[4:36])
+
+		// initialize new block info
+		newNode := &BlockNode{
+			Height:     height,
+			Hash:       hash,
+			ParentHash: parentHash,
+			Children:   []*BlockNode{},
+		}
+
+		// map block info corresponding to its hash
+		blocks[hash] = newNode
+
+		// condition to append block info to existing blockInfos according to their block height
+		if _, found := deepReorgs[height]; found {
+			// add block address as its parent's children
+			blocks[newNode.ParentHash].Children = append(blocks[newNode.ParentHash].Children, newNode)
+			deepReorgs[height] = append(deepReorgs[height], newNode)
+		} else {
+			if isFirst {
+				isFirst = false
+			} else {
+				// add block address as its parent's children
+				blocks[newNode.ParentHash].Children = append(blocks[newNode.ParentHash].Children, newNode)
```

Comment:
What if `newNode.ParentHash` is not present? I am getting nil dereference

Reply:
Before attempting to append to the Children slice, we an check if blocks[newNode.ParentHash] exists
```
if parent, ok := blocks[newNode.ParentHash]; ok {
	parent.Children = append(parent.Children, newNode)
} else {
	// Handle the case where the parent doesn't exist
       // skip
}
```
If the block data is coming in sequential order (from smallest to largest block height), then we would always have the parent block before any of its children. If you cannot guarantee this, if we can not guarantee this then we should place the above check